pkg/statistics/handle/autoanalyze/priorityqueue/queue_v2.go

```diff
@@ -0,0 +1,247 @@
+	}
+	pq.inner = heap.NewHeap(keyFunc, lessFunc)
+	if err := pq.init(); err != nil {
+		pq.Close()
+		return errors.Trace(err)
+	}
+
+	// Start a goroutine to maintain the priority queue.
+	pq.wg.Run(pq.run)
+	pq.initialized.Store(true)
+	return nil
+}
+
+// init initializes the priority queue and adds jobs to it.
+func (pq *AnalysisPriorityQueueV2) init() error {
+	return statsutil.CallWithSCtx(pq.statsHandle.SPool(), func(sctx sessionctx.Context) error {
+		parameters := exec.GetAutoAnalyzeParameters(sctx)
+		err := pq.setAutoAnalysisTimeWindow(parameters)
+		if err != nil {
+			return err
+		}
+		if !pq.IsWithinTimeWindow() {
+			return nil
+		}
+		timeWindow := pq.autoAnalysisTimeWindow.Load().(*AutoAnalysisTimeWindow)
+
+		err = FetchAllTablesAndBuildAnalysisJobs(
+			sctx,
+			parameters,
+			*timeWindow,
+			pq.statsHandle,
+			pq.Push,
+		)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		return nil
+	}, statsutil.FlagWrapTxn)
+}
+
+// run maintains the priority queue.
+func (pq *AnalysisPriorityQueueV2) run() {
+	defer func() {
+		if r := recover(); r != nil {
+			statslogutil.StatsLogger().Error("Priority queue panicked", zap.Any("recover", r), zap.Stack("stack"))
+		}
+	}()
+
+	timeRefreshInterval := time.NewTicker(lastAnalysisDurationRefreshInterval)
+	defer timeRefreshInterval.Stop()
+
+	for {
+		select {
+		case <-pq.ctx.Done():
+			statslogutil.StatsLogger().Info("Priority queue stopped")
+			return
+		case <-timeRefreshInterval.C:
+			statslogutil.StatsLogger().Info("Start to refresh last analysis durations of jobs")
+			pq.RefreshLastAnalysisDuration()
+		}
+	}
+}
+
+// RefreshLastAnalysisDuration refreshes the last analysis duration of all jobs in the priority queue.
+func (pq *AnalysisPriorityQueueV2) RefreshLastAnalysisDuration() {
+	if !pq.IsWithinTimeWindow() {
+		statslogutil.StatsLogger().Debug("Not within the auto analyze time window, skip refreshing last analysis duration")
+		return
+	}
+
+	start := time.Now()
+	defer func() {
+		statslogutil.StatsLogger().Info("Last analysis duration refreshed", zap.Duration("duration", time.Since(start)))
+	}()
+	if err := statsutil.CallWithSCtx(pq.statsHandle.SPool(), func(sctx sessionctx.Context) error {
+		jobs := pq.inner.List()
+		for _, job := range jobs {
+			indicators := job.GetIndicators()
+			currentTs, err := getStartTs(sctx)
+			if err != nil {
+				return errors.Trace(err)
+			}
+
+			jobFactory := NewAnalysisJobFactory(sctx, 0, currentTs)
+			tableStats, ok := pq.statsHandle.Get(job.GetTableID())
+			if !ok {
+				statslogutil.StatsLogger().Warn("Table stats not found during refreshing last analysis duration",
+					zap.Int64("tableID", job.GetTableID()),
+					zap.String("job", job.String()),
+				)
+				continue
+			}
+			indicators.LastAnalysisDuration = jobFactory.GetTableLastAnalyzeDuration(tableStats)
+			job.SetIndicators(indicators)
+			job.SetWeight(pq.calculator.CalculateWeight(job))
+			if err := pq.inner.Add(job); err != nil {
```

Comment:
Ohhh, I just find that the Add behaves as adding or in-place updating.
Should we change the API for a better understanding?
My original question is why we are adding it here since the `List` will not clean the map.

Reply:
It has another API called `AddIfNotPresent`. I guess I should call `Update` here.

Reply:
Updated.